### PR TITLE
Fix for http error 404: unable to login

### PR DESCRIPTION
**Pull request by goegol**
Fix for http error 404: unable to login by mail
Login, receive the mail, accept it within 2 minutes and you are logged in!

### CODE REVIEW

File: tgtg/__init__.py [hunk 1]
```diff
@@ -207,15 +247,15 @@ def set_favorite(self, item_id, is_favorite):
...
+            email,
+            password,
+            name,
+            country_id="GB",
+            device_type="ANDROID",
+            newsletter_opt_in=False,
+            push_notification_opt_in=True,
     ):
         response = requests.post(
             self._get_url(SIGNUP_BY_EMAIL_ENDPOINT),
```
Comment (DL6ER) [inline tgtg/__init__.py:261]: You removed `SIGNUP_BY_EMAIL_ENDPOINT` above

**Review by DL6ER — COMMENTED** [review]
Are all the indentation changes intended?

File: tgtg/__init__.py [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
 
 import requests
 
-from .exceptions import TgtgAPIError, TgtgLoginError
+from .exceptions import TgtgAPIError, TgtgLoginError, TgtgPollingError
 
 BASE_URL = "https://apptoogoodtogo.com/api/"
 API_ITEM_ENDPOINT = "item/v7/"
 LOGIN_ENDPOINT = "auth/v2/loginByEmail"
-SIGNUP_BY_EMAIL_ENDPOINT = "auth/v2/signUpByEmail"
```
Comment (DL6ER) [inline tgtg/__init__.py:13]: Removed here

File: tgtg/__init__.py [hunk 3]
```diff
@@ -1,16 +1,18 @@
...
+AUTH_BY_EMAIL_ENDPOINT = "auth/v3/authByEmail"
```
Comment (DL6ER) [inline tgtg/__init__.py:14]:
```suggestion
SIGNUP_BY_EMAIL_ENDPOINT = "auth/v2/signUpByEmail"
AUTH_BY_EMAIL_ENDPOINT = "auth/v3/authByEmail"
```

File: tgtg/__init__.py (hunk 1 above)
Comment (Dielee) [reply tgtg/__init__.py:261]: Yes, right. This is the new tgtg login endpoint.

File: tgtg/__init__.py (hunk 3 above)
Comment (Dielee) [reply tgtg/__init__.py:14]: Why this endpoint should stay in this lib ? It's not used anymore.

Comment (DL6ER) [reply tgtg/__init__.py:14]: It is currently still used in `signup_by_email`. If you remove it there, it can be removed here, too.

Comment (Dielee) [reply tgtg/__init__.py:14]: Ah yes, your right. 
But this function is never used, so it's garbage. I will remove it at the afternoon.

Comment (Dielee) [reply tgtg/__init__.py:14]: Fixed!

**Comment by Dielee** [comment]
@ahivert will you merge my changes ?

**Comment by ahivert** [comment]
> @ahivert will you merge my changes ?

Sorry missed that PR, can you please fix indentation with black ?

This project use pre-commit inside the project you have to run:
```
pre-commit install
pre-commit run -a
```

File: tgtg/__init__.py [hunk 4]
```diff
@@ -96,57 +100,97 @@ def _refresh_token(self):
...
+
             else:
-                raise TgtgLoginError(response.status_code, response.content)
+                if response.status_code == 429:
+                    raise TgtgAPIError("429 - Too many requests. Try again later.")
+                else:
+                    raise TgtgLoginError(response.status_code, response.content)
+
+    def start_polling(self, polling_id):
+        polling_response_code = 202
```
Comment (ahivert) [inline tgtg/__init__.py:140]:
```suggestion
        polling_response_code = HTTPStatus.ACCEPTED
```

File: tgtg/__init__.py [hunk 5]
```diff
@@ -96,57 +100,97 @@ def _refresh_token(self):
...
+        i = 1
+
+        while polling_response_code == 202 and i <= MAX_POLLING_TRIES:
```
Comment (ahivert) [inline tgtg/__init__.py:143]: can you please use a forloop here ?
something like:
```
for retry_index in range(MAX_POLLING_TRIES):
```

File: tgtg/__init__.py [hunk 6]
```diff
@@ -96,57 +100,97 @@ def _refresh_token(self):
...
+                json={
+                    "device_type": "ANDROID",
+                    "email": self.email,
+                    "request_polling_id": polling_id
+                },
+                proxies=self.proxies,
+                timeout=self.timeout,
+            )
+            polling_response_code = response.status_code
+            if polling_response_code == 202 or polling_response_code == 200:
```
Comment (ahivert) [inline tgtg/__init__.py:156]:
```suggestion
            if polling_response_code in (HTTPStatus.ACCEPTED, HTTPStatus.OK):
```

File: tgtg/__init__.py [hunk 7]
```diff
@@ -96,57 +100,97 @@ def _refresh_token(self):
...
+                if response.status_code == 200:
```
Comment (ahivert) [inline tgtg/__init__.py:157]:
```suggestion
                if response.status_code == HTTPStatus.OK:
```

File: tgtg/__init__.py [hunk 8]
```diff
@@ -19,22 +22,23 @@
     "TGTG/21.9.3 Dalvik/2.1.0 (Linux; Android 6.0.1; SM-G920V Build/MMB29K)",
 ]
 DEFAULT_ACCESS_TOKEN_LIFETIME = 3600 * 4  # 4 hours
+MAX_POLLING_TRIES = 24  # 24 * 5 seconds = 2 minutes
```
Comment (ahivert) [inline tgtg/__init__.py:25]: We need another constant to define delay (5s) between retries

**Review by ahivert — CHANGES_REQUESTED** [review]

**Comment by ahivert** [comment]
@Dielee thanks for this PR ! Just some remarks to fix and please check that tests still works. 
I will not ask you to add some tests (I will do it later) but keep existing tests green :)

File: tgtg/__init__.py (hunk 5 above)
Comment (Dielee) [reply tgtg/__init__.py]: Fixed!

File: tgtg/__init__.py (hunk 8 above)
Comment (Dielee) [reply tgtg/__init__.py]: Added!

**Comment by Dielee** [comment]
@ahivert changed the requested changes, fixed black, but don't know how to fix the isort error...

**Comment by ahivert** [comment]
thanks @Dielee I will continue the work on #125 ;)

⚠️ CI on master is failing so please do not use master branch if you don't know what you do

**Pull request merged**